### PR TITLE
Timeout

### PR DESCRIPTION
**Pull request by SRFU-NN**
Adding timeout to requests
Alerting user of slow steps, i.e. where the default timeout has been changed.

### CODE REVIEW

File: src/OptiHPLCHandler/empower_api_core.py [hunk 1]
```diff
@@ -115,47 +133,85 @@ def logout(self) -> None:
...
+    ) -> requests.Response:
         """
-        Get data from Empower.
+        Wrapper for requests.
 
-        :param endpoint: The endpoint to get data from.
+        :param method: The method to use.
+        :param endpoint: The endpoint to use.
+        :param body: The body to use.
         """
```
Comment (Firefly78) [inline src/OptiHPLCHandler/empower_api_core.py:145]: MIssing parameter in doc-string.

File: src/OptiHPLCHandler/empower_api_core.py [hunk 2]
```diff
@@ -115,47 +133,85 @@ def logout(self) -> None:
...
+
+        def _request_with_timeout(method, endpoint, header, body, timeout):
+            try:
+                response = requests.request(
```
Comment (Firefly78) [inline src/OptiHPLCHandler/empower_api_core.py:149]: Just return the result straight away.

File: src/OptiHPLCHandler/empower_api_core.py [hunk 3]
```diff
@@ -115,47 +133,85 @@ def logout(self) -> None:
...
+                    method,
+                    endpoint,
+                    json=body,
+                    headers=header,
+                    timeout=timeout,
+                )
+            except requests.exceptions.Timeout as e:
+                print(f"{method}ing {body} to {endpoint} timed out")
```
Comment (Firefly78) [inline src/OptiHPLCHandler/empower_api_core.py:157]: Again with defining the same string multiple times.

File: src/OptiHPLCHandler/empower_api_core.py [hunk 4]
```diff
@@ -115,47 +133,85 @@ def logout(self) -> None:
...
+                logger.error("%sing %s to %s timed out", method, body, endpoint)
+                raise Exception(f"{method}ing {body} to {endpoint} timed out") from e
```
Comment (Firefly78) [inline src/OptiHPLCHandler/empower_api_core.py:159]: Raise timeout exception?

**Review by Firefly78 — CHANGES_REQUESTED** [review]

File: src/OptiHPLCHandler/empower_api_core.py [hunk 5]
```diff
@@ -50,7 +51,14 @@ def __init__(
...
             logger.debug("No service specified, getting service from Empower")
-            response = requests.get(self.address + "/authentication/db-service-list")
+            try:
+                response = requests.get(
+                    self.address + "/authentication/db-service-list", timeout=10
+                )
+            except requests.exceptions.Timeout as e:
+                print(f"Getting service from {self.address} timed out")
+                logger.error("Getting service from %s timed out", self.address)
+                raise Exception(f"Getting service from {self.address} timed out") from e
```
[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:61]: Will do.

File: src/OptiHPLCHandler/empower_api_core.py [hunk 6]
```diff
@@ -50,7 +51,14 @@ def __init__(
         self.username = getpass.getuser()
         if service is None:
             logger.debug("No service specified, getting service from Empower")
-            response = requests.get(self.address + "/authentication/db-service-list")
+            try:
+                response = requests.get(
+                    self.address + "/authentication/db-service-list", timeout=10
+                )
+            except requests.exceptions.Timeout as e:
+                print(f"Getting service from {self.address} timed out")
```
[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:59]: Makes sense, will do.

File: src/OptiHPLCHandler/empower_api_core.py [hunk 7]
```diff
@@ -87,10 +95,20 @@ def login(
...
-            json=body,
-        )
+        try:
+            reply = requests.post(
+                self.address + "/authentication/login",
+                json=body,
+                timeout=60,
+            )
+        except requests.exceptions.Timeout as e:
+            print(f"Login to {self.address} with username = {self.username} timed out")
```
[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:105]: Fixed

File: src/OptiHPLCHandler/empower_api_core.py [hunk 8]
```diff
@@ -87,10 +95,20 @@ def login(
...
+            logger.error(
+                "Login to %s with username = %s timed out", self.address, self.username
+            )
+            raise Exception(
```
[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:109]: Fixed

File: src/OptiHPLCHandler/empower_api_core.py (hunk 2 above)
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:149]: Hmm, that feels dirty, somehow. But it also seems like a better way of doing it.

File: src/OptiHPLCHandler/empower_api_core.py (hunk 1 above)
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:145]: Another question is whether to even have a docstring, it is an internal function. But now it is there, so I might as well keep it.

File: src/OptiHPLCHandler/empower_api_core.py (hunk 3 above)
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:157]: fixed

File: src/OptiHPLCHandler/empower_api_core.py (hunk 4 above)
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:159]: Fixed.

File: src/OptiHPLCHandler/empower_api_core.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py:61]: This is a different exception from what the requests one. I assume that is not a problem?

[the remark this answers is not in the document]
Comment (Firefly78) [reply src/OptiHPLCHandler/empower_api_core.py]: Just with a more clear message.

[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py]: I can't figure out which is better. What would you do?

[the remark this answers is not in the document]
Comment (Firefly78) [reply src/OptiHPLCHandler/empower_api_core.py]: I think (and SQBL agree) that the exception should be of type "requests.exceptions.Timeout".

[the remark this answers is not in the document]
Comment (SRFU-NN) [reply src/OptiHPLCHandler/empower_api_core.py]: Fixed.

**Review by Firefly78 — APPROVED** [review]
All good.

**Pull request merged**